File: examples/python/08_fallback_policy.py
import audio_sdk


def main() -> None:
    discovery_manager = audio_sdk.AudioManager()
    devices = discovery_manager.enumerate_devices()
    fallback_input = audio_sdk.first_device(devices, audio_sdk.DeviceDirection.INPUT)
    if fallback_input is None:
        raise SystemExit("No available input device for fallback policy example")

    config = audio_sdk.AudioConfig()
    config.input.preferred_id = "missing-input-device-id"
    config.input.fallback_id = fallback_input.id

    manager = audio_sdk.AudioManager(config)

    def on_event(event: object) -> None:
        print(
            f"event={audio_sdk.event_type_name(event.type)} "
            f"device={event.device_id} "
            f'details="{event.details}"'
        )

    manager.set_event_callback(on_event)

    path = audio_sdk.recording_path("python_fallback_policy_example.wav")
    print(
        "Preferred input is intentionally missing; fallback should be "
        f"{fallback_input.name} ({fallback_input.id})"
    )

    audio_sdk.require(manager.start_recording(str(path)), "start fallback recording")
    audio_sdk.sleep_for_recording()
    audio_sdk.require(manager.stop_recording(), "stop fallback recording")
    print(f"Fallback recording completed at {path}")


if __name__ == "__main__":
    main()

File: examples/cpp/08_fallback_policy.cpp
#include <iostream>

#include "example_utils.h"

int main() {
    audio_sdk::AudioManager discovery_manager;
    const auto devices = discovery_manager.EnumerateDevices();
    const auto fallback_input = audio_sdk_examples::FirstDevice(devices, audio_sdk::DeviceDirection::kInput);

    if (!fallback_input) {
        std::cerr << "No available input device for fallback policy example\n";
        return 1;
    }

    audio_sdk::AudioConfig config;
    config.input.preferred_id = "missing-input-device-id";
    config.input.fallback_id = fallback_input->id;

    audio_sdk::AudioManager manager(config);
    manager.SetEventCallback([](const audio_sdk::Event& event) {
        std::cout << "event=" << audio_sdk_examples::EventTypeName(event.type)
                  << " device=" << event.device_id
                  << " details=\"" << event.details << "\"\n";
    });

    const std::filesystem::path path = audio_sdk_examples::RecordingPath("cpp_fallback_policy_example.wav");
    std::cout << "Preferred input is intentionally missing; fallback should be "
              << fallback_input->name << " (" << fallback_input->id << ")\n";

    if (!audio_sdk_examples::Require(manager.StartRecording(path.string()), "start fallback recording")) {
        return 1;
    }

    audio_sdk_examples::SleepForRecording();
    if (!audio_sdk_examples::Require(manager.StopRecording(), "stop fallback recording")) {
        return 1;
    }

    std::cout << "Fallback recording completed at " << path << '\n';
    return 0;
}
